gson/src/main/java/com/google/gson/stream/JsonWriter.java

```diff
@@ -162,6 +167,190 @@
     HTML_SAFE_REPLACEMENT_CHARS['\''] = "\\u0027";
   }
 
+  private class StringValWriter extends Writer {
+    private boolean isClosed;
+
+    private void verifyNotClosed() throws IOException {
+      if (isClosed) {
+        throw new IOException("Writer is closed");
+      }
+    }
+
+    @Override
+    public void write(int c) throws IOException {
+      verifyNotClosed();
+      stringPiece((char) c);
+    }
+
+    /**
+     * @param len
+     *      Length of the data, e.g. array length
+     * @param off
+     *      0-based offset where the section begins
+     * @param sectLen
+     *      Length of the section
+     * @throws IndexOutOfBoundsException
+     *      If {@code off} or {@code sectLen} is invalid
+     */
+    private void validateIndices(int len, int off, int sectLen) {
+      if (off < 0) {
+        throw new IndexOutOfBoundsException("offset < 0");
+      } else if (sectLen < 0) {
+        throw new IndexOutOfBoundsException("length < 0");
+      } else if (sectLen > len - off) {
+        throw new IndexOutOfBoundsException("length > data.length - offset");
+      }
+    }
+
+    @Override
+    public void write(char[] cbuf, int off, int len) throws IOException {
+      verifyNotClosed();
+      validateIndices(cbuf.length, off, len);
+      stringPiece(cbuf, off, len);
+    }
+
+    @Override
+    public void write(String str, int off, int len) throws IOException {
+      verifyNotClosed();
+      validateIndices(str.length(), off, len);
+      stringPiece(str.substring(off, off + len));
+    }
+
+    @Override
+    public void write(String str) throws IOException {
+      verifyNotClosed();
```

Comment:
This doesn't look thread safe but might be OK if these objects never escape one thread.

Reply:
The doc for JsonWriter says:
> Instances of this class are not thread safe.

I hope that should prevent users from passing the created writers to different threads.